lib/logstash/outputs/google_bigquery.rb

```diff
@@ -141,539 +130,151 @@ class LogStash::Outputs::GoogleBigQuery < LogStash::Outputs::Base
-  config :flush_interval_secs, :validate => :number, :default => 2
+  # If logstash is running within Google Compute Engine, the plugin will use
+  # GCE's Application Default Credentials. Outside of GCE, you will need to
+  # specify a Service Account JSON key file.
+  config :json_key_file, validate: :string, required: false
 
-  # Uploader interval when uploading new files to BigQuery. Adjust time based
-  # on your time pattern (for example, for hourly files, this interval can be
-  # around one hour).
-  config :uploader_interval_secs, :validate => :number, :default => 60
+  # The number of messages to upload at a single time. (< 1000, default: 128)
+  config :batch_size, validate: :number, required: true, default: 128
 
-  # Deleter interval when checking if upload jobs are done for file deletion.
-  # This only affects how long files are on the hard disk after the job is done.
-  config :deleter_interval_secs, :validate => :number, :default => 60
+  # An approximate number of bytes to upload as part of a batch. Default: 1MB
+  config :batch_size_bytes, validate: :number, required: true, default: 1_000_000
 
-  public
-  def register
-    require 'csv'
-    require "fileutils"
-    require "thread"
+  # Uploads all data this often even if other upload criteria aren't met. Default: 5s
+  config :flush_interval_secs, validate: :number, required: true, default: 5
 
-    @logger.debug("BQ: register plugin")
-
-    if !@csv_schema.nil?
-      @fields = Array.new
-
-      CSV.parse(@csv_schema.gsub('\"', '""')).flatten.each do |field|
-        temp = field.strip.split(":")
-
-        # Check that the field in the schema follows the format (<name>:<value>)
-        if temp.length != 2
-          raise "BigQuery schema must follow the format <field-name>:<field-value>"
-        end
+  # The location to store events that could not be uploaded due to errors.
+  # Consider using an additional Logstash input to pipe the contents of
+  # these to an alert platform so you can manually fix the events.
+  #
+  # Or use https://cloud.google.com/storage/docs/gcs-fuse[GCS FUSE] to
+  # transparently upload to a GCS bucket.
+  #
+  # Files names follow the pattern `[table name]-[UNIX timestamp].log`
+  config :error_directory, validate: :string, required: true, default: '/tmp/bigquery_errors'
+
+  # The following configuration options still exist to alert users that are using them
+  config :uploader_interval_secs, validate: :number, obsolete: 'No longer used.'
+  config :deleter_interval_secs, validate: :number, obsolete: 'No longer used.'
+  config :key_path, validate: :string, obsolete: 'Use json_key_file or ADC instead.'
+  config :key_password, validate: :string, obsolete: 'Use json_key_file or ADC instead.'
+  config :service_account, validate: :string, obsolete: 'Use json_key_file or ADC instead.'
+  config :temp_file_prefix, validate: :string, obsolete: 'No longer used.'
+  config :temp_directory, validate: :string, obsolete: 'No longer used.'
 
```

Comment:
Would you remind deprecating rather than obsoleting the 'no longer used' settings - I assume from your comments in the changelog that an existing pipeline using the `ADC` will continue to work with this implementation? Deprecating the settings enables the pipeline to continue working with warnings, until the config setting is marked `obsolete` at a later date, when using that setting will fail the pipeline.

Reply:
Sure thing! I'll mark all  of them except the `key_path` and `key_password`. If someone was using ADC those shouldn't be defined at all.

If they were using the old key style, I'd rather that blow up on them than silently falling back to credentials that might be sitting in their environment variables that they explicitly overrode with the P12 style key file. How does that sound?

Reply:
That makes sense - thanks!